### PR TITLE
[OpenWRT] Update the Makefile and init scripts

### PR DESCRIPTION
**Pull request by caipiblack**
Some modifications to make `ot-br-posix` working with OpenWRT.

Note that it is also compatible with Matter (tested on TE9)

This pull request is related to this issue: https://github.com/openthread/ot-br-posix/issues/1250

### CODE REVIEW

**Comment by caipiblack** [comment]
Using ot-br-posix on OpenWRT
----------------------------

This document describe how to setup `ot-br-posix` on OpenWRT. For the example we are using a `WRT3200ACM` router.

## Install OpenWRT on WRT3200ACM

Go to the OpenWRT wiki page about this router and execute the instructions. [click here](https://openwrt.org/toh/linksys/wrt3200acm)

This example use the release `21.02.3` but you can use a more recent.

## SDK Setup

1. Go to the download page of OpenWRT for the target: `https://downloads.openwrt.org/releases/21.02.3/targets/mvebu/cortexa9/`.
2. Download the SDK `openwrt-sdk-21.02.3-mvebu-cortexa9_gcc-8.4.0_musl_eabi.Linux-x86_64.tar.xz`.
3. Extract the archive into a folder `sdk`.
4. Copy `feeds.conf.default` to `feeds.conf`.
5. Add a new line in `feeds.conf`: `src-link openthread /media/mathieu/data/Projects/ot-br-posix/etc/openwrt`.
6. Update the feeds and install `openthread-br`:
```
./scripts/feeds update -a
./scripts/feeds install openthread-br
./scripts/feeds install iptables-mod-extra ipset
```
_NOTE: `ipset` is required for the firewall feature and `iptables-mod-extra` is required to get the iptable match `pkttype`._
7. Configure the SDK (optional):
```
make menuconfig
# Disable: Global build settings > Cryptographically sign package lists
# Disable: Advanced configuration options (for developers) > Automatic rebuild of packages
# Disable: Advanced configuration options (for developers) > Automatic removal of build directories
```
8. Build:
```
make defconfig
make -j
```
9. On the OpenWRT router, create a temporary folder to store the generated packages: `mkdir /tmp/pkg`.
10. Transfer the required packages to the router:
```
scp \
bin/packages/arm_cortex-a9_vfpv3-d16/openthread/openthread-br_1.0-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libdbus_1.13.12-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/dbus_1.13.12-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libdaemon_0.14-5_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libexpat_2.2.10-2_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libavahi-dbus-support_0.8-7_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/avahi-dbus-daemon_0.8-7_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/packages/libavahi-client_0.8-7_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/base/ipset_7.6-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/base/libmnl0_1.0.4-2_arm_cortex-a9_vfpv3-d16.ipk \
bin/packages/arm_cortex-a9_vfpv3-d16/base/libipset13_7.6-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/librt_1.1.24-3_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/libstdcpp6_8.4.0-3_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/iptables-mod-extra_1.8.7-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-usb-acm_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-tun_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-ipt-extra_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-ipt-ipset_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
bin/targets/mvebu/cortexa9/packages/kmod-nfnetlink_5.4.188-1_arm_cortex-a9_vfpv3-d16.ipk \
root@192.168.2.1:/tmp/pkg
```
11. Install the packages:

```
opkg install /tmp/pkg/*
```

The `otbr-agent` and `otbr-firewall` serviced are automatically started after install and when the system boot.

## Test with Thread client

1. On the router, create the thread network:
```
ot-ctl dataset init new
ot-ctl dataset commit active
ot-ctl prefix add fd11:22::/64 pasor
ot-ctl ifconfig up
ot-ctl thread start
ot-ctl netdata register
ot-ctl dataset active -x
```
2. On a thread device with `ot-cli` firmware, Configure it to join the thread network:
```
dataset set active 0e080000000000010000000300001335060004001fffe00208528351a158b36e550708fd372f9a2edde39e0510730eeb4d8bcf632e2b17c46329f0b394030f4f70656e5468726561642d616662630102afbc0410e5c6daa09b2eb053fa563b6a3db002b00c0402a0f7f8
ifconfig up
thread start
ipaddr
```
3. Make sure you can ping the `fd11:` IP of the thread device from the router.
4. Make sure you can ping the `fd11:` IP of the router from the thread device.
5. Make sure you can ping the `fd11:` IP of the thread device from your computer (the computer must be connected to a LAN port of the router).

_In case of failure at the point #5, use wireshark to check if your computer receive a ICMPv6 message with router advertisement when the thread network is created and make sure your computer accept it._

## Test with Matter device

For the test I am using a `ǹRF52840 DK` with `lighting-app` example on `TE9` tag.

1. Build and flash the firmware on the device according the the README of the example [here](https://github.com/project-chip/connectedhomeip/tree/master/examples/lighting-app/nrfconnect).
2. Commission the device using `chip-tool`:
```
./chip-tool pairing qrcode-thread 10 hex:0e080000000000010000000300001335060004001fffe00208528351a158b36e550708fd372f9a2edde39e0510730eeb4d8bcf632e2b17c46329f0b394030f4f70656e5468726561642d616662630102afbc0410e5c6daa09b2eb053fa563b6a3db002b00c0402a0f7f8 MT:6FCJ142C00KA0648G00
```
3. Control the device using `chip-tool`:
```
./chip-tool onoff toggle 10 1
```

File: src/openwrt/otbr-agent.init.in [hunk 1]
```diff
@@ -34,6 +34,6 @@ USE_PROCD=1
 
 start_service() {
 	procd_open_instance
-	procd_set_param command @CMAKE_INSTALL_FULL_SBINDIR@/otbr-agent 'spinel+hdlc+uart:///dev/ttyUSB0?uart-baudrate=115200'
+	procd_set_param command @CMAKE_INSTALL_FULL_SBINDIR@/otbr-agent -I wpan0 -B @OTBR_INFRA_IF_NAME@ spinel+hdlc+uart:///dev/ttyACM0 trel://@OTBR_INFRA_IF_NAME@
```
Comment (bukepo) [inline src/openwrt/otbr-agent.init.in:37]: Ideally we should make this device configurable with uci.

**Review by bukepo — APPROVED** [review]
Thanks!

**Comment by caipiblack** [comment]
## Update

- As suggested by @bukepo : The INFRA_IF_NAME is now configurable by uci
- I also made configurable the THREAD_IF_NAME by uci as it is used by two scripts (`/etc/init.d/otbr-agent` and `/etc/init.d/otbr-firewall`) 

**Notes:**

- At this point it is also possible to make the RADIO_URL configurable through uci.
- The file `ot-br-posix/src/openwrt/otbr-agent.init.in` has executable flag. It is not realy required because during IPK creation the  executable flag is added to the file. This is made by the makefile (`ot-br-posix/etc/openwrt/openthread-br/Makefile`) when we do `$(INSTALL_BIN) (...the file...) (...destination...)`. I don't change it as I don't know if it's wanted or not. If it's wanted, then `otbr-firewall.init.in` should have the same flag.
- It is also possible to make all uci parameters configurable by web page. (need some work and UI design)

## This is how to interact with uci

```
root@OpenWrt:~# uci show
...
otbr-agent.service=otbr-agent
otbr-agent.service.infra_if_name='br-lan'
otbr-agent.service.thread_if_name='wpan0'
...
root@OpenWrt:~# service otbr-firewall stop
root@OpenWrt:~# service otbr-agent stop
root@OpenWrt:~# uci set otbr-agent.service.thread_if_name=wpan1
root@OpenWrt:~# uci commit otbr-agent
root@OpenWrt:~# service otbr-firewall start
root@OpenWrt:~# service otbr-agent start
```
**Notes:**

- When uci settings are changed, `otbr-agent` and `otbr-firewall` must be restarted for the new settings to take effect.
- If you change `otbr-agent.service.thread_if_name` in uci while `otbr-firewall` is in execution, then when the process `otbr-firewall` is stopped it will not be able to cleanup the iptables rules. So it's important to stop the service before changing the uci settings or just reboot. (iptables are cleaned on each boot)

## global notes

-- Currently the make-pretty return errors in lot of files that are not impacted by this PR.

**Comment by LorbusChris** [comment]
Apologies for the bump. Are there any blockers for these changes?

There is an ongoing effort to get otbr properly added to the OpenWRT distro: https://github.com/openwrt/packages/pull/19218
It'd be great to get this PR in, so the downstream package author can rebase their package onto these changes before it lands in OpenWRT.

**Comment by caipiblack** [comment]
Currently with my changes there is still some points to check:

- I made INFRA_IF_NAME configurable through UCI based on @bukepo suggestion. And I think it should be nice now if it’s also configurable from web ui. But i don’t really know where and how (need ui design ideas/instructions or someone to do it)
- I also made configurable THREAD_IF_NAME but i’m not sure if it’s really good because of firewall script issue* and also because in the web ui there are references to “wpan0” statically, so it’s probably not supposed to be configurable. (In such case, I can revert changes about THREAD_IF_NAME). And we can do the same remark as before, it should be good if it’s also configurable from web page..

(*)  The issue with firewall script:

> If you change otbr-agent.service.thread_if_name in uci while otbr-firewall is in execution, then when the process otbr-firewall is stopped it will not be able to cleanup the iptables rules. So it's important to stop the service before changing the uci settings or just reboot. (iptables are cleaned on each boot)

I don’t really know if it’s a problem;

- The firewall script is supposed to be launched when the system starts to configure iptables. It’s not really supposed to be stopped (like otbr-agent service)
- But, if we want there are solutions to fix it, for exemple when the firewall script is started we get the THREAD_IF_NAME info from uci, then we set firewall based on this interface name and we store the interface name in /tmp/(some_file). And when the service is stopped, it will simply read the THREAD_IF_NAME from the file in /tmp instead of uci. This is to prevent problems if user change THREAD_IF_NAME from uci while openthread is running.


I’m going to fix the issue of firewall script soon, but i need some ideas for the points that i mentioned before.

**Comment by caipiblack** [comment]
Note that it is also possible to put the content of otbr-firewall script in otbr-agent startup script.

i use two scripts just to keep the same philosophy as what is made for ubuntu/debian

**Comment by caipiblack** [comment]
At this point:

- INFRA_IF_NAME is configurable with UCI
- THREAD_IF_NAME is configurable with UCI
- The script otbr-firewall is fixed (in case of THREAD_IF_NAME changes while the script is running, we can still stop it and remove ipset rules)

Remaining work:

- (optionnal) : Make INFRA_IF_NAME and THREAD_IF_NAME also configurable from the web page. (need ui design ideas/instructions or someone to do it)
- In the web page there are references to "wpan0" statically (`src/openwrt/view/admin_thread/thread_overview.htm`), it should be good to display the correct interface name, depending on the interface used by otbr-agent. This point is not necessary related to this PR.

**Pull request closed without merging**